src/coreclr/jit/llvm.cpp

```diff
@@ -159,6 +163,70 @@ unsigned Llvm::getElementSize(CORINFO_CLASS_HANDLE classHandle, CorInfoType corI
     return getWellKnownTypeSize(corInfoType);
 }
 
+StructDesc* Llvm::getStructDesc(CORINFO_CLASS_HANDLE structHandle)
+{
+    if (_structDescMap->find(structHandle) == _structDescMap->end())
+    {
+        TypeDescriptor structTypeDescriptor = _getTypeDescriptor(_thisPtr, structHandle);
+        unsigned structSize                 = _info.compCompHnd->getClassSize(structHandle); // TODO-LLVM: add to TypeDescriptor?
+
+        std::vector<CORINFO_FIELD_HANDLE> sparseFields = std::vector<CORINFO_FIELD_HANDLE>(structSize);
+
+        for (unsigned i = 0; i < structSize; i++)
+            sparseFields[i] = nullptr;
+
+        // determine the largest field for unions, and get fields in order of offset
+        for (unsigned i = 0; i < structTypeDescriptor.fieldCount; i++)
+        {
+            CORINFO_FIELD_HANDLE fieldHandle = structTypeDescriptor.fields[i];
+            unsigned fldOffset               = _info.compCompHnd->getFieldOffset(structTypeDescriptor.fields[i]);
+
+            assert(fldOffset < structSize);
+
+            unsigned fieldSize = _info.compCompHnd->getClassSize(_info.compCompHnd->getFieldClass(fieldHandle));
+            // store the biggest field at the offset for unions
+            if (sparseFields[fldOffset] == nullptr ||
+                fieldSize > _info.compCompHnd->getClassSize(_info.compCompHnd->getFieldClass(sparseFields[fldOffset])))
+            {
+                sparseFields[fldOffset] = fieldHandle;
+            }
+        }
+
+        // count the struct fields after replacing fields with equal offsets
+        unsigned fieldCount = 0;
+        for (unsigned i = 0; i < structSize; i++)
+        {
+            if (sparseFields[i] != nullptr)
+            {
+                fieldCount++;
+            }
+        }
+
+        StructDesc* structDesc = new StructDesc(fieldCount, new FieldDesc[fieldCount]);
+
+        unsigned fieldIx = 0;
+        for (unsigned fldOffset = 0; fldOffset < structSize; fldOffset++)
+        {
+            if (sparseFields[fldOffset] == nullptr)
+            {
+                continue;
+            }
+
+            FieldDesc* fieldDesc = structDesc->getFieldDesc(fieldIx);
+
+            CORINFO_FIELD_HANDLE fieldHandle = structTypeDescriptor.fields[fieldIx];
+            CORINFO_CLASS_HANDLE fieldClassHandle = NO_CLASS_HANDLE;
+
+            fieldDesc->setFieldData(_info.compCompHnd->getFieldOffset(fieldHandle),
+                                    _info.compCompHnd->getFieldType(fieldHandle, &fieldClassHandle), fieldClassHandle);
+            fieldIx++;
+        }
```

Comment:
It took some time waiting around in the debugger, but I did eventually find the cause behind the OOM issue...

Reproduction:
```cs
[MethodImpl(MethodImplOptions.NoInlining)]
private static void Problem(DateTime v1, DateTimeOffset v2, DateTime* p1, DateTimeOffset* p2)
{
    *p1 = v1;
    *p2 = v2;
}
```
This code is using `structTypeDescriptor.fields[fieldIx]` to get the field handle, but at this point, the fields may have been reversed (such as the case for `DateTimeOffset` auto-layout struct), it should be using `sparseFields[fldOffset] instead.

Also, seem like the `getFieldOffset` call is not necessary (we already have `fldOffset`)?

Also, I believe it is the case in C++ that the `getFieldType`'s result has to be into its own local for this to have well-defined behavior, otherwise the compiler can reorder the arguments such that `fieldClassHandle` is evaluated before `getFieldType` has been called.

Reply:
Thanks for looking at this.  And confirm it allows the tests to run in parallel again.  See what the CI makes of the release build now.  The LLVM was missing all the structs in the RyuJIT module so not sure it is related.